### PR TITLE
Add React and basic video streaming features

### PR DESCRIPTION
**Pull request by kamil-stasiak**
## What works:
- camera and screen-sharing streaming
- camera and screen-sharing playing
- start / stop camera button
- start / stop screen-sharing button
- prettier

## Not yet implemented:
- audio streaming 
- audio playing
- permission manager - better info for user
- leave button (for now there is only an icon)
- socket / webrtc communication and error handling
- select room screen (for now it always connects to room ID: "123" and every browser tab gets a new UUID)

## Needs work:
<details>
  <summary>css grid</summary>
  
image
</details>
- remove console.logs
- configs (tailwind, eslint, prettier, esbuild, etc.)
- eslint doesn't work yet

## Features added only for development
<details>
  <summary>screen-sharing icon on peer tile</summary>
  
  image
</details>
<details>
  <summary>random emoji icon as a user metadata </summary>
  
  image
</details>
<details>
  <summary>camera streaming indicator</summary>
  
  image
</details>

### CODE REVIEW

**Comment by kamil-stasiak** [comment]
> * nice use of hooks 👍

Thank you 😊

> * do you think it would be wise to have membrane hooks as a library for others to use? we have membrane for React Native (it has even a similar api to your hooks), wouldn't it be a good idea to have membrane for React as well if we're writing hooks etc anyway?

I think it would be nice to extract hooks as a library. We have to have a look also at 'membrne live' project. 

> * are you going to write the whole frontend in one PR? this would be a nightmare to review 😅 and recently we had a talk about dividing up tasks

This is the first step. I will add new commits as PR to this PR. Then I will merge everything to the main branch. 

> * is it normal in web dev that you have 200+ character long lines with styles o.o? sometimes partially copied? seems not readable to me (I code in React Native and I've never used tailwind before, just normal styles and styled-components)

It's tailwind way. It's strange but it works.

> * why is camera off by default ;_;

Only for development. It's easier to test. W will turn camera on by default when I implement permission manager.

> * in some places (when you have a complicated object in the state and you're updating state based on previous value, for example in `useMedia`) maybe it'll be worth using `useReducer` (https://reactjs.org/docs/hooks-reference.html#usereducer)?

This is nice! I forgot about It!

Thank you!

File: assets/src/pages/room/hooks/usePeerState.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  const addTrack = (
+    peerId: string,
+    trackId: string,
+    mediaStreamTrack?: MediaStreamTrack,
+    mediaStream?: MediaStream,
+    metadata?: Metadata
+  ) => {
+    setPeers((prev: Peers) => {
+      const peerCopy: LocalPeer = { ...prev[peerId] };
+      const oldTracks: Track[] = peerCopy.tracks.filter((e) => e.trackId !== trackId);
```
Comment (Rados13) [inline assets/src/pages/room/hooks/usePeerState.ts:65]: This line is used here and in removeTrack. What do you think about extracting it to function?

**Review by Rados13 — COMMENTED** [review]
The whole PR looks good. Sorry if I asked about something that's not been implemented yet, but I decided that's better to ask than forget about it 😅 .

**Comment by bblaszkow06** [comment]
@kamil-stasiak please rename the branch to the proper prefix (Rewrite is MV-87, not MV-82)

File: assets/src/pages/room/components/MediaControlButtons.tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+const getControls = (
+  userMediaAudio: UseMediaResult,
+  userMediaVideo: UseMediaResult,
+  displayMedia: UseMediaResult,
+  navigate: NavigateFunction
+) => [
+  userMediaAudio.stream
+    ? {
+        icon: "/svg/mic-line.svg",
+        hover: "Mute or unmute the microphone",
```
Comment (bblaszkow06) [inline assets/src/pages/room/components/MediaControlButtons.tsx:16]:
```suggestion
        hover: "Mute the microphone",
```

File: assets/src/pages/room/components/MediaControlButtons.tsx [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        onClick: () => {
+          console.log("Stopping mic...");
+          userMediaAudio.stop();
+        },
+      }
+    : {
+        icon: "/svg/mic-off-fill.svg",
+        hover: "Mute or unmute the microphone",
```
Comment (bblaszkow06) [inline assets/src/pages/room/components/MediaControlButtons.tsx:24]:
```suggestion
        hover: "Unmute the microphone",
```

File: assets/src/pages/room/components/MediaControlButtons.tsx [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        onClick: () => {
+          console.log("Starting mic...");
+          userMediaAudio.start();
+        },
+      },
+  userMediaVideo.stream
+    ? {
+        icon: "/svg/camera-line.svg",
+        hover: "Turn on/off the camera",
```
Comment (bblaszkow06) [inline assets/src/pages/room/components/MediaControlButtons.tsx:33]:
```suggestion
        hover: "Turn off the camera",
```

File: assets/src/pages/room/components/MediaControlButtons.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        onClick: () => {
+          console.log("Stopping camera...");
+          userMediaVideo.stop();
+          // todo remove stream
+        },
+      }
+    : {
+        hover: "Turn on/off the camera",
+        icon: "/svg/camera-off-line.svg",
```
Comment (bblaszkow06) [inline assets/src/pages/room/components/MediaControlButtons.tsx:42]:
```suggestion
        icon: "/svg/camera-off-line.svg",
        hover: "Turn on the camera",
```

File: assets/src/pages/room/components/MediaControlButtons.tsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        icon: "/svg/computer-line.svg",
+        hover: "Start or stop the screensharing",
+        onClick: () => {
+          console.log("Stopping ScreenSharing...");
+          displayMedia.stop();
+        },
+      }
+    : {
+        icon: "/svg/computer-line.svg",
+        hover: "Start or stop the screensharing",
```
Comment (bblaszkow06) [inline assets/src/pages/room/components/MediaControlButtons.tsx:59]:
```suggestion
        hover: "Start the screensharing",
```

File: assets/src/pages/room/components/MediaControlButtons.tsx [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        onClick: () => {
+          console.log("Starting ScreenSharing...");
+          displayMedia.start();
+        },
+      },
+  {
+    icon: "/svg/phone-fill.svg",
+    hover: "Leave button",
```
Comment (bblaszkow06) [inline assets/src/pages/room/components/MediaControlButtons.tsx:67]:
```suggestion
    hover: "Leave the room",
```

File: assets/src/pages/room/components/VideoPlayer.tsx [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      >
+        Camera turned off
+      </div>
+    </div>
+  );
+};
+
+// 1. autoplay onLoad - play() failed when video is not muted and user didn't interact with the page
+// from chrome console: play() failed because the user didn't interact with the document first. https://goo.gl/xX8pDD
+// https://stackoverflow.com/questions/49930680/how-to-handle-uncaught-in-promise-domexception-play-failed-because-the-use
```
Comment (bblaszkow06) [inline assets/src/pages/room/components/VideoPlayer.tsx:72]: IIRC calling one of the WebRTC API functions (successful `getUserMedia` I think) also may allow enabling autoplay

File: assets/src/old/room.ts [hunk 9]
```diff
@@ -235,9 +237,7 @@ export class Room {
     }
 
     try {
-      this.localAudioStream = await navigator.mediaDevices.getUserMedia({
-        audio: AUDIO_TRACK_CONSTRAINTS,
-      });
+      this.localAudioStream = await navigator.mediaDevices.getUserMedia();
```
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply assets/src/old/room.ts:240]: It's back:
https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-1f29e8da7ded7f83effca3899fc0dbd3f618ee2f45e80492697f33a736d49b33

File: assets/src/pages/room/components/VideoPeerPlayers.tsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+const getCameraStreams = (
+  peers: Peers,
+  videoStream?: MediaStream,
+  audioStream?: MediaStream
+): MediaStreamWithMetadata[] =>
+  Object.values(peers).map((peer) => {
+    const video: Track | undefined = peer.tracks.find((track) => track?.metadata?.type === "camera");
+    const screenSharingStream: MediaStream | undefined = peer.tracks.find(
+      (track) => track?.metadata?.type === "screensharing"
+    )?.mediaStream;
```
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply assets/src/pages/room/components/VideoPeerPlayers.tsx:23]: Fixed:
https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-75624838b56d1124ca6b954c8344eeca923b08e0dc5a35134f198632722d2995R28-R38

File: assets/src/pages/room/components/VideoPlayer.tsx [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+      {metadata?.bottomLeft && (
+          <div data-name="video-label" className="absolute text-white text-shadow-lg bottom-0 left-0 p-2">
+            {metadata.bottomLeft}
+          </div>
+      )}
+      {metadata?.bottomRight && (
+          <div data-name="video-label" className="absolute text-white text-shadow-lg bottom-0 right-0 p-2">
+            {metadata.bottomRight}
+          </div>
+      )}
```
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply assets/src/pages/room/components/VideoPlayer.tsx:58]: Fixed:
https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-f35a113f5e34466e1af4290a524acbbbda2528b372659eae52cdfd25434453a9

File: assets/src/pages/room/hooks/usePeerState.ts [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  peers: Peers;
+  addPeers: (peerId: string[]) => void;
+  removePeer: (peerId: string) => void;
+  addTrack: (
+    peerId: string,
+    trackId: string,
+    mediaStreamTrack?: MediaStreamTrack,
+    mediaStream?: MediaStream,
+    metadata?: Metadata
+  ) => void;
```
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply assets/src/pages/room/hooks/usePeerState.ts:34]: Fixed: 
https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-0c557d56c602c5197821fe45e65ff491805c28e979e3d0d756465bf64cf3eb4cR26

File: assets/src/pages/room/hooks/usePeerState.ts [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+  mediaStreamTrack?: MediaStreamTrack;
+  mediaStream?: MediaStream;
+  metadata?: Metadata;
+};
+
+export type LocalPeer = {
+  id: string;
+  tracks: Track[];
+  removedTracks: string[];
+};
```
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply assets/src/pages/room/hooks/usePeerState.ts:14]: Yes, there it is:
https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-30ac30cb732aa7c1e426057d17a692d96d9f3ccdef852773c9f028fb3dd5acd4R40

File: assets/src/pages/room/hooks/usePeerState.ts [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+  ) => void;
+  removeTrack: (peerId: string, trackId: string) => void;
+};
+
+export function usePeersState(): UsePeersStateResult {
+  const [peers, setPeers] = useState<Peers>({});
+
+  const addPeers = (peerIds: string[]) => {
+    setPeers((prevState: Peers) => {
+      const newPeers: Peers = Object.fromEntries(peerIds.map((id) => [id, { id: id, tracks: [], removedTracks: [] }]));
```
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply assets/src/pages/room/hooks/usePeerState.ts:43]: Great idea. There is a new implementation:
https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-30ac30cb732aa7c1e426057d17a692d96d9f3ccdef852773c9f028fb3dd5acd4R38-R47

File: assets/src/pages/room/components/MediaControlButtons.tsx (hunk 2 above)
Comment (kamil-stasiak) [reply assets/src/pages/room/components/MediaControlButtons.tsx:16]: Fixed:
https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-ab2b4039f281ecf95bf165270892779885a477b31b4642ec14ce0613a8b7d288

File: assets/src/pages/room/components/MediaControlButtons.tsx (hunk 3 above)
Comment (kamil-stasiak) [reply assets/src/pages/room/components/MediaControlButtons.tsx:24]: Fixed: https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-ab2b4039f281ecf95bf165270892779885a477b31b4642ec14ce0613a8b7d288

File: assets/src/pages/room/components/MediaControlButtons.tsx (hunk 4 above)
Comment (kamil-stasiak) [reply assets/src/pages/room/components/MediaControlButtons.tsx:33]: Fixed: https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-ab2b4039f281ecf95bf165270892779885a477b31b4642ec14ce0613a8b7d288

File: assets/src/pages/room/components/MediaControlButtons.tsx [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+        icon: "/svg/camera-off-line.svg",
+        onClick: () => {
+          console.log("Starting camera...");
+          userMediaVideo.start();
+        },
+      },
+  displayMedia.stream
+    ? {
+        icon: "/svg/computer-line.svg",
+        hover: "Start or stop the screensharing",
```
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply assets/src/pages/room/components/MediaControlButtons.tsx:51]: Fixed: https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-ab2b4039f281ecf95bf165270892779885a477b31b4642ec14ce0613a8b7d288

File: assets/src/pages/room/components/MediaControlButtons.tsx (hunk 6 above)
Comment (kamil-stasiak) [reply assets/src/pages/room/components/MediaControlButtons.tsx:59]: Fixed: https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-ab2b4039f281ecf95bf165270892779885a477b31b4642ec14ce0613a8b7d288

File: assets/src/pages/room/components/MediaControlButtons.tsx (hunk 7 above)
Comment (kamil-stasiak) [reply assets/src/pages/room/components/MediaControlButtons.tsx:67]: Fixed: https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-ab2b4039f281ecf95bf165270892779885a477b31b4642ec14ce0613a8b7d288

File: assets/src/pages/room/components/VideoPlayer.tsx (hunk 8 above)
Comment (kamil-stasiak) [reply assets/src/pages/room/components/VideoPlayer.tsx:72]: I removed this comment. This was a note for me at the begging. Everything is working so I decided to delete it

**Comment by kamil-stasiak** [comment]
> @kamil-stasiak please rename the branch to the proper prefix (Rewrite is MV-87, not MV-82)

Final merge will be from branch with right name.

File: assets/src/pages/room/components/ScreenSharingPlayers.tsx [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+  videoStream?: MediaStream;
+};
+
+const prepareScreenSharingStreams = (
+  peers: Peers,
+  localStream?: MediaStream
+): { screenSharingStreams: VideoStreamWithMetadata[]; isScreenSharingActive: boolean } => {
+  const peersScreenSharingTracks: VideoStreamWithMetadata[] = Object.values(peers)
+    .flatMap((peer) => peer.tracks.map((track) => ({ peerId: peer.id, track: track })))
+    .filter((e) => e.track?.metadata?.type === "screensharing")
```
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply assets/src/pages/room/components/ScreenSharingPlayers.tsx:17]: Fixed

File: assets/src/pages/room/hooks/useMembraneClient.tsx [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+  removePeer: (peerId: string) => void,
+  addTrack: (
+    peerId: string,
+    trackId: string,
+    mediaStreamTrack?: MediaStreamTrack,
+    mediaStream?: MediaStream,
+    metadata?: Metadata
+  ) => void,
+  removeTrack: (peerId: string, trackId: string) => void,
+  setErrorMessage: (value: ((prevState: string | undefined) => string | undefined) | string | undefined) => void
```
[the remark this answers is not in the document]
Comment (kamil-stasiak) [reply assets/src/pages/room/hooks/useMembraneClient.tsx:35]: Extracted: 
https://github.com/membraneframework/membrane_videoroom/pull/61/files#diff-229d084a72e521648558e39521ecd68bb6b2f20870f40e7f7bba747819bff13aR20

**Pull request closed without merging**